[ML] Refactor data extraction logic and improve cleanup handling

valeriy42 opened the pull request:
Fixes scroll context leaks in datafeed CCS tests by ensuring proper cleanup of scroll contexts.

Changes include:
- Wrapping DataExtractor usage in a try-finally block within `DatafeedJob.run()` to guarantee that `destroy()` is always invoked.
- Calling `clearScrollLoggingExceptions()` in `ScrollDataExtractor.clearScroll()` to handle failures smoothly.
- Destroys the previous extractor in `ChunkedDataExtractor.advanceTime()` before instantiating a new one.
- Adds a cleanup wait in the test after stopping the datafeed.
- Removes the `@AwaitsFix` annotation from `DatafeedCcsIT`.

Fixes https://github.com/elastic/elasticsearch/issues/84268

valeriy42 committed:
Refactor data extraction logic and improve cleanup handling

- Removed the outdated @AwaitsFix annotation from DatafeedCcsIT.
- Enhanced the data extraction process in DatafeedJob by restructuring the while loop for better readability and error handling.
- Added cleanup logic to ensure scroll contexts are properly destroyed after data extraction in both DatafeedJob and ChunkedDataExtractor.
- Updated ScrollDataExtractor to improve error handling during scroll clearing.

These changes aim to improve the robustness and maintainability of the data extraction process in the ML module.

valeriy42 commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DatafeedJob.java:
```diff
@@ -363,112 +363,117 @@ private void run(long start, long end, FlushJobAction.Request flushRequest) {
+            }
+        } finally {
+            // Ensure the extractor is always destroyed to clean up scroll contexts
+            dataExtractor.destroy();
```
This is the block I actually added

elasticsearchmachine commented:
Pinging @elastic/ml-core (Team:ML)

jan-elastic reviewed (approved):
LGTM. Just one thing that can hopefully be improved

valeriy42 committed:
Add busy wait instead of waiting for context baseline.

valeriy42 committed:
declare lambda method throw exception

valeriy42 committed:
Enhance DatafeedCcsIT cleanup logic

- Added a sleep period after stopping the datafeed and job to ensure scroll requests complete, particularly following network disruptions.
- Increased the timeout for context checks from 5 to 30 seconds to improve reliability in waiting for expected active states.

valeriy42 committed:
Refine DatafeedCcsIT cleanup and context handling

- Removed the sleep period after stopping the datafeed and job, as the cleanup mechanism should now handle scroll requests effectively.
- Updated documentation to clarify the cleanup process and its reliance on the datafeed's mechanisms.
- Increased the timeout for context checks from 30 to 60 seconds to enhance reliability after network recovery.

valeriy42 committed:
Improve error handling in DatafeedCcsIT during context cleanup

- Added exception handling when waiting for contexts to return to baseline to ensure cleanup proceeds even if the wait fails.
- Updated assertions to check for datafeed errors more clearly, enhancing the robustness of the test logic.
- Improved logging to capture failures in context recovery, aiding in debugging and test reliability.

valeriy42 committed:
Merge branch 'main' into tests/is-84268
